### PR TITLE
Add comprehensive input validation with detailed error responses

### PR DESCRIPTION
**Pull request by olaservo**
## Summary
- Added comprehensive input validation middleware for all Todo endpoints
- Implemented field-specific error messages with structured error responses  
- Added validation for title length, HTML tag prevention, and type checking
- Enhanced ID parameter validation for route security
- Added comprehensive test coverage for all validation scenarios

### New Files
- `src/middleware/validation.ts` - Core validation logic with custom error handling
- `src/middleware/validation.test.ts` - Comprehensive test suite (15+ test cases)

### Modified Files
- `src/routes/todos.ts` - Integrated validation middleware, improved error responses
- `README.md` - Added validation rules and error response format documentation
- `package.json` - Added supertest for API testing

### Validation Rules Implemented
- **Title**: Required, 1-100 chars, no HTML tags, string type validation
- **Description**: Optional, max 500 chars, no HTML tags, string type validation  
- **Completed**: Boolean type validation when provided
- **ID Parameters**: Positive integer validation for all routes

## Test Coverage
- Valid input acceptance
- Missing required fields
- Invalid data types
- Length limit enforcement
- HTML tag prevention (XSS protection)
- Multiple simultaneous errors
- Edge cases and boundary conditions

## API Response Changes
**Before**: Simple error strings
```json
{ "error": "Title is required" }
```

**After**: Structured error objects with field details
```json
{
  "error": "Validation failed", 
  "details": [
    { "field": "title", "message": "Title is required" },
    { "field": "description", "message": "Description must be 500 characters or less" }
  ]
}
```

## Security Improvements
- HTML tag prevention to mitigate XSS attacks
- Input sanitization with automatic trimming
- Type validation to prevent injection attacks
- ID format validation to prevent path traversal

## Testing Instructions

1. **Install dependencies**: `npm install`
2. **Run test suite**: `npm test` (should show 15+ passing tests)
3. **Test API manually**:
   ```bash
   npm run dev
   
   # Test valid creation
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"Valid Todo","description":"Valid description"}'
   
   # Test validation errors
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"","description":"'$(printf 'a%.0s' {1..501})'"}'
   ```

## Breaking Changes
**Error response format has changed** - API consumers will need to handle the new structured error format

## Future Enhancements
- [ ] Add rate limiting validation
- [ ] Implement custom validation rules configuration
- [ ] Add internationalization for error messages
- [ ] Consider adding request sanitization middleware

### CODE REVIEW

File: src/middleware/validation.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    errors.push({ field: 'title', message: 'Title cannot contain HTML tags' });
+  }
+
+  // Validate description
+  if (description !== undefined && description !== null) {
+    if (typeof description !== 'string') {
+      errors.push({ field: 'description', message: 'Description must be a string', value: description });
+    } else if (description.length > 500) {
+      errors.push({ field: 'description', message: 'Description must be 500 characters or less' });
+    } else if (/<.*>/g.test(description)) {  // BUG: Inconsistent regex - allows <img> but title rejects it
```
Comment (dandeliongold) [inline src/middleware/validation.ts:42]: **Bug: Inconsistent HTML tag validation regex**

The regex for description HTML validation `/<.*>/g` is different from the title validation `/<[^>]*>/g`. The description regex will match `` but the title regex won't, creating inconsistent behavior.

**Recommendation:** Use the same regex pattern for both:
```typescript
} else if (/<[^>]*>/g.test(description)) {
```

File: src/middleware/validation.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+export const validateTodoId = (req: Request, res: Response, next: NextFunction) => {
+  const { id } = req.params;
+
+  if (!id || !/^\d+$/.test(id)) {
+    return res.status(400).json({
+      error: 'Invalid ID format',
+      details: [{ field: 'id', message: 'ID must be a positive integer', value: id }]
+    });
+  }
+
```
Comment (dandeliongold) [inline src/middleware/validation.ts:109]: **Bug: ID "0" edge case**

The current validation allows ID "0" which may not be valid depending on your ID generation strategy. Most systems start IDs from 1.

**Recommendation:** Consider adding a check for positive integers:
```typescript
if (!id || !/^\d+$/.test(id) || parseInt(id) === 0) {
```

File: src/routes/todos.ts [hunk 3]
```diff
@@ -9,54 +10,67 @@ router.get('/', (req: Request, res: Response) => {
...
+router.post('/', validateTodoCreate, (req: Request, res: Response) => {
   const { title, description }: CreateTodoRequest = req.body;
   
-  if (!title || title.trim() === '') {
-    return res.status(400).json({ error: 'Title is required' });
-  }
-  
-  const todo = todoStore.create({ title: title.trim(), description });
+  // BUG: No try-catch block - if todoStore.create throws, server will crash
+  const todo = todoStore.create({
```
Comment (dandeliongold) [inline src/routes/todos.ts:31]: **Bug: Missing error handling**

If `todoStore.create` throws an exception, the server will crash since there's no try-catch block.

**Recommendation:** Add error handling:
```typescript
try {
  const todo = todoStore.create({ 
    title: title.trim(), 
    description: description?.trim() || undefined 
  });
  res.status(201).json(todo);
} catch (error) {
  res.status(500).json({ 
    error: 'Internal server error',
    details: [{ field: 'server', message: 'Failed to create todo' }]
  });
}
```

File: src/middleware/validation.test.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        .post('/')
+        .send({ title: '<script>alert("xss")</script>' });
+
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'title',
+        message: 'Title cannot contain HTML tags'
+      });
+    });
+
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:61]: **Missing test case**

You're missing a test case for description HTML tag validation. The comment indicates this is a known gap.

**Recommendation:** Add this test:
```typescript
it('should reject description with HTML tags', async () => {
  const response = await request(app)
    .post('/')
    .send({ title: 'Valid Title', description: '<script>alert("xss")</script>' });

  expect(response.status).toBe(400);
  expect(response.body.details).toContainEqual({
    field: 'description',
    message: 'Description cannot contain HTML tags'
  });
});
```

File: src/middleware/validation.test.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    it('should reject non-numeric ID', async () => {
+      const response = await request(app).get('/abc');
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'id',
+        message: 'ID must be a positive integer',
+        value: 'abc'
+      });
+    });
+
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:119]: **Missing test case for ID edge case**

As mentioned in the comment, you should test the ID "0" edge case.

**Recommendation:** Add this test:
```typescript
it('should reject ID zero', async () => {
  const response = await request(app).get('/0');
  expect(response.status).toBe(400);
  expect(response.body.details).toContainEqual({
    field: 'id',
    message: 'ID must be a positive integer',
    value: '0'
  });
});
```

**Review by dandeliongold — CHANGES_REQUESTED** [review]
## Overall Assessment

This is a **well-structured PR** that significantly improves the Todo API with comprehensive input validation and security enhancements. The implementation is largely solid, but there are several critical bugs and missing test cases that need attention.

## ✅ **Strengths**

1. **Comprehensive validation coverage** - Validates title, description, completed status, and ID parameters
2. **Security-focused** - HTML tag prevention helps mitigate XSS attacks
3. **Excellent error structure** - The new structured error format is much better than simple strings
4. **Good test coverage** - 15+ test cases cover most validation scenarios
5. **Clear documentation** - Well-documented validation rules and error format in README
6. **Breaking change clarity** - Properly documented that error response format has changed

## 🚨 **Critical Issues to Fix**

### 1. **HTML Validation Bug** (High Priority)
- Inconsistent regex patterns between title and description validation
- Could allow XSS vectors in descriptions that would be blocked in titles

### 2. **Server Crash Risk** (High Priority)  
- No error handling in POST route - server will crash if `todoStore.create` throws
- Should wrap in try-catch block

### 3. **ID Validation Gap** (Medium Priority)
- Allows ID "0" which may not be valid depending on ID generation strategy
- Most systems start IDs from 1

### 4. **Missing Test Coverage** (Medium Priority)
- Missing test for description HTML validation
- Missing test for ID "0" edge case
- Both are noted in comments but not implemented

## 🔄 **Recommendations**

1. **Fix the HTML regex inconsistency** - Use same pattern for both fields
2. **Add error handling** to prevent server crashes
3. **Validate positive integers only** for IDs (exclude 0)
4. **Complete the test suite** - Add the missing test cases noted in comments

## 💡 **Nice to Have**

- Consider adding input sanitization middleware
- Rate limiting validation (as mentioned in future enhancements)
- Internationalization support for error messages

The foundation is excellent, but please address the critical bugs before merging to ensure production stability and security.
